### ISSUE TITLE
Copy all contract data to another contract; join N contracts into one

### ISSUE THREAD

**Issue by vporton** [ethereum/EIPs#3032]
Need a new kind of operation (both as a new RPC command and a new opcode) that creates a contract from a given bytecode _and_ copies (to implement it efficiently use copy on write) all data from another contract.

This is to allow forking erroneous or otherwise disadvantageous contracts. It is better than upgradeable contracts:

- works for every contract (non-upgradeable contracts, too)
- has no overhead of indirect calls
- unlike upgradeable contracts, does not cancel the old contract, what is important for trust and stability, for decentralization of management of contract logic

Moreover, we could join forked data from several contracts (specified when doing the fork operation) 0 .. N, with a new Solidity keyword `contract_data` (and new opcodes) like:

```solidity
SolidityContract0 data0 = SolidityContract0(new contract_data(0));
SolidityContract1 data1 = SolidityContract1(new contract_data(1));
...
SolidityContractN dataN = SolidityContractN(new contract_data(N));
...
data0.variable = 123;
```

That is we would allow forking several contracts 0..N into one contract. It should be optional to designate one contract K of the contracts 0..N as the main contract (so that `variable` would become equivalent to `dataK.variable`). No need for designation of this contract K on EVM level, it can be done by an option to Solidity compiler or a compiler directive.

The feature to join N contracts into one is not strictly necessary, as can be done using several wrapper contracts each of which forks just one of the N contract, but allowing to join N contracts into one is a performance advantage as it does not require deployment of N new contracts and wrappers to access their data.

**Comment by vporton** [comment]
Hm, we could also make `new contract_data` to accept an address as an argument and allow to fork contracts dynamically without creating a new contract. Not sure if passing address to `new contract_data` is really worth, but I am inclined that implementing this would be worth, too.

**Comment by vporton** [comment]
Oh, no need for `new contract_data(NUMBER)` syntax and no need to specify a list of N contracts on a new contract creating. We can do it instead just with `new contract_data(ADDRESS)` operation!

**Comment by MicahZoltu** [comment]
I believe something like this has unbounded costs if naively implemented, and it isn't immediately obvious how one would implement it in a non-naive way.  Would such a non-naive implementation work in both the go-ethereum/Nethermind storage model and the TurboGeth storage model?

**Comment by vporton** [comment]
> I believe something like this has unbounded costs if naively implemented

Constant-time implementation:

Keep an array of arrays. The first index is the contract + index of the next contract in this array (linked list), the second index is the storage slot. The values are values of storage slots.

Keep an array of first indexes in the above array indexed by storage slots.

To access a value in a storage slot read the first array, then index the second array by the value obtained from the first array and the given storage slot.

When a variable value is changed, the value in the second array changes to the index of the very first contract.

Please check this algorithm for bugs.

Really constant? Well, I suspect memory allocations may make it non-constant.

But the requirement to copy contract data "manually" one-by-one item using much gas and also breaking atomicity is _much_ worse in costs than implementing the above algorithm.

**Comment by MicahZoltu** [comment]
All contract storage in Ethereum is in a Patricia Merkle trie, so unfortunately it isn't as simple as it would be in a traditional application.  Contract data isn't (and can't be) stored in an array, just due to the nature of data storage and consensus and proofs.

**Comment by bigcook12345** [comment]
So tell me why, with everything going on, I'm blocked from everything that
makes obviously my respitories work!?!? Did I go straight to you and snatch
it right out of your hands, or did I pull request!? The whole
"Stop-opt-out" thing sure as heck threw a wrench in the work I was trying
to do since day one. What part of "PULL-REQUEST", doesn't self explain!?

All of this is to keep pulling a request and figure out what works and what
doesn't. Am I wrong..?

Well it might have pissed a people off. But that's just it!? I wasn't here
to do that either!?

Because when I get lost in something, I try what I need to, to make it
right/work. Well if you've never ran into any walls while working, then by
all means, "Kudos". I want to be just like you when I grow up 🤦‍♂️. Either
way, it wasn't to uniform for you guy's. But guess what.

My running into walls helped you find your next way of correcting what was
wrong in the 1st place. I just don't mind saying that it me 🤷‍♂️. And look
how much it helped...

**Comment by vporton** [comment]
> All contract storage in Ethereum is in a Patricia Merkle trie, so unfortunately it isn't as simple as it would be in a traditional application. Contract data isn't (and can't be) stored in an array, just due to the nature of data storage and consensus and proofs.

I don't know Ethereum internals, but it is weird for me that storing data in Merkle trie is necessary: We need to verify _changes_ (e.g. the contract bytecodes and function arguments) match between nodes, not the current state.

Isn't it possible to rewrite the entire Ethereum database without that Merkle trie? I realize the change is major, but that's an important feature.

**Comment by MicahZoltu** [comment]
A Merkle Patricia Trie (MPT) is necessary to ensure that state is the same for all clients in the network, and to enable things like light clients, stateless clients, data proofs, etc.  The MPT design to storage is a pretty core part of the Ethereum protocol, so much so that changing it significantly is a *major* undertaking.  TurboGeth is currently working on a way to store the trie on disk in a different format, while still being able to achieve all of the guarantees, in order to conserve disk space and improve performance and this task is a multi-year project for a full team.  Removal of the MPT would probably be the largest possible change to Ethereum, and realistically a change of that magnitude would be much better suited for a separate blockchain, just do to how deeply rooted MPTs are in the design/structure of modern blockchains.

**Comment by vporton** [comment]
> A Merkle Patricia Trie (MPT) is necessary to ensure that state is the same for all clients in the network, and to enable things like light clients, stateless clients, data proofs, etc. The MPT design to storage is a pretty core part of the Ethereum protocol, so much so that changing it significantly is a _major_ undertaking. TurboGeth is currently working on a way to store the trie on disk in a different format, while still being able to achieve all of the guarantees, in order to conserve disk space and improve performance and this task is a multi-year project for a full team. Removal of the MPT would probably be the largest possible change to Ethereum, and realistically a change of that magnitude would be much better suited for a separate blockchain, just do to how deeply rooted MPTs are in the design/structure of modern blockchains.

But I do not understand why it is necessary to store the entire data in MPT: Why not just store in the blockchain every change (amounts of transactions, hashes and arguments of called functions)? The rest could be able to be restored from it. Isn't Ethereum deterministic?

I think this would _much_ improve the performance.

Sorry, if I say stupid things.

**Comment by vporton** [comment]
I mean to store on disk of a node changes in a blockchain format and store the current states (values of storage slots, etc.) in a non-cryptographic format.

**Comment by vporton** [comment]
Oh, I've got: It is to enable light clients.

I propose a possibly minor change in the Ethereum standard: Make MPT-related features optional. So not every node would have support for light clients. Then we would develop "non-verified light clients" that would just trust the node.

**Comment by vporton** [comment]
After a little more thinking:

I am not a cryptography expert, but I assume that storing my "arrays of arrays" (see above) in MPT instead or in addition of what is currently stored in MPT will _not_ cause unbounded costs (if the current MPT storage model does not cause unbounded costs, because my model of storage seems then similar to the current).

It possibly requires a change in protocol: To verify light clients need to use _another_ MPT. But this change is very useful. Somebody said it needs another blockchain. In my opinion, it should be done as an Ethereum hardfork (if it really requires a hardfork).

**Comment by MicahZoltu** [comment]
If you really want to go down this path, my recommendation would be to start by forking one of the Ethereum clients out there and modifying it to work the way you desire.  This will have one of two possible results:
1. You figure out a clever solution to the problem that allows the clients to continue to function as they are while gaining new functionality.
2. You gain a better understanding of why this is a particularly hard problem.

The TurboGeth team is exercising this strategy for a separate thing and they are finding (1) to be the case.  They are sharing their results and it is promising for everyone.  However, many other people have gone with this strategy and ended up in case (2), which at least has really helped those people understand the architecture and often they go on to make great contributions in other areas, even though their original goal was never realized.

Re: Light clients.  There is a strong desire to *improve* light client support in Ethereum, so anything that doesn't help with that will likely see a lot of negative feedback.

**Comment by vporton** [comment]
I made a child error in my "solution". The "second array" is pertinent to each contract, so it needs to be copied every time a contract is forked. So it is not O(1) as I claimed.

However, there [exists](https://en.wikipedia.org/w/index.php?title=Persistent_array&oldid=969141238#Log-log-time) a  O(log log(min(n,m)) solution where m is the number of forks and n is the total number of used storage slots.

log log(min(n,m)) is not so bad by itself, but it creates a problem: The cost of gas would need to be adjusted as the data structure grows to keep the complexity constant, otherwise complexity per a wei of gas grows unbounded (however slow).

Is it OK to allow the gas complexity to grow slowly? I think, it's OK: The cost of the CPUs decreases faster than that's inverse. And we eliminate giant gas usage needed to copy a contract data to other contract "manually".
